### ISSUE TITLE
AR Model training discussion on custom prefix to dub audio.

### ISSUE THREAD

**Issue by kdcyberdude** [Plachtaa/VALL-E-X#116]
I am currently in the process of modifying the Auto Regressive model code to dub audio into a different language. 

To illustrate, the input would be _English_Output_Text + Hindi_Audio_Prefix_, and the output would be _English_Audio_.  Where I want the number of quantized frames of _English_Audio_ to match _Hindi_Audio_Prefix_. 

I have made the following changes - Here **en_audio_features** is English_Audio and **hi_audio_features** is Hindi_Audio_Prefix.
The fundamental idea in the following code is that, I have utilized the combined text and Hindi_Audio_prefix in **'x'** and generated sequence audio i.e. English_Audio as **'y'**. Accordingly, I have adjusted the attention mask.
```python
# NOTE: x has been padded in TextTokenCollater
        x_mask = make_pad_mask(x_lens).to(x.device)
        y_mask = make_pad_mask(y_lens).to(en_audio_features.device)
        y_mask_int = y_mask.type(torch.int64)

        text = x
        codes = en_audio_features.type(torch.int64) * (1 - y_mask_int.unsqueeze(dim=-1))
        codes_prefix = hi_audio_features.type(torch.int64) * (1 - y_mask_int.unsqueeze(dim=-1))

        y, targets = self.pad_y_eos(
            codes[..., 0], y_mask_int, eos_id=NUM_AUDIO_TOKENS
        )

        y_prefix, targets_prefix = self.pad_y_eos(
            codes_prefix[..., 0], y_mask_int, eos_id=NUM_AUDIO_TOKENS
        )

        x_len = x_lens.max()

        metrics = {}
        total_loss = 0.0

        xy_padding_mask = torch.concat([x_mask, y_mask, y_mask], dim=1)
        if self.ar_audio_prepend_bos:
            ar_xy_padding_mask = torch.concat(
                [x_mask, F.pad(y_mask, (1, 0), value=False), F.pad(y_mask, (1, 0), value=False)], dim=1
            )
        else:
            ar_xy_padding_mask = xy_padding_mask
        # AR Decoder
        if train_stage in [0, 1]:
            x = self.ar_text_embedding(text)
            x = self.ar_text_prenet(x)
            x = self.ar_text_position(x)

            y_len = y_lens.max() + int(self.ar_audio_prepend_bos)

            x_attn_mask = F.pad(
                torch.zeros((x_len + y_len, x_len + y_len), dtype=torch.bool, device=x.device),
                (0, y_len),
                value=True,
            )

            y_attn_mask = F.pad(
                torch.triu(
                    torch.ones(y_len, y_len, dtype=torch.bool, device=x.device),
                    diagonal=1,
                ),
                (x_len + y_len, 0),
                value=False,
            )
            xy_attn_mask = torch.concat([x_attn_mask, y_attn_mask], dim=0)

            # merge key padding and attention masks
            bsz, src_len = x.shape[0], x_len + y_len + y_len
            _xy_padding_mask = (
                ar_xy_padding_mask.view(bsz, 1, 1, src_len)
                .expand(-1, self.num_heads, -1, -1)
                .reshape(bsz * self.num_heads, 1, src_len)
            )
            xy_attn_mask = xy_attn_mask.logical_or(_xy_padding_mask)

            new_attn_mask = torch.zeros_like(xy_attn_mask, dtype=x.dtype)
            new_attn_mask.masked_fill_(xy_attn_mask, float("-inf"))
            xy_attn_mask = new_attn_mask

            y = torch.concat([y_prefix, y], dim=1)

            y_emb = self.ar_audio_embedding(y)
            y_emb = self.ar_audio_prenet(y_emb)
            y_pos = self.ar_audio_position(y_emb)

            xy_pos = torch.concat([x, y_pos], dim=1)

            xy_dec, _ = self.ar_decoder(
                (xy_pos, None),
                mask=xy_attn_mask,
                # src_key_padding_mask=xy_padding_mask,
                # is_causal=True,
            )
            logits = self.ar_predict_layer(xy_dec[:, x_len + y_len:]).permute(0, 2, 1)
            # loss
            total_loss = F.cross_entropy(logits, targets, reduction=reduction)

            metrics["ArTop10Accuracy"] = self.ar_accuracy_metric(
                logits.detach(), targets
            ).item() * y_lens.sum().type(torch.float32)

        if self.num_quantizers == 1:
            return ((x, codes), total_loss, metrics)

        # Non-AR Decoders
```

I am eager to discuss and brainstorm about the approach. Does the aforementioned method seem viable, or should we consider an alternative strategy? Since I am new to this I'll really appreciate the help!!

**Comment by kdcyberdude** [comment]
Hi @Plachtaa, M4T has the potential to be used for dubbing, but it would require architectural changes to incorporate both text and speech as controllable parameters.

The VALL-E-X architecture already has Text(TTS) and Audio prompt(Reference speech/prefix) as controllable parameters to generate the speech. This architecture seems like a suitable choice. Furthermore, in addition to TTS, my aim is to generate speech with the same characteristics (tone, feeling, style, etc.) as the reference speech.

I have created a dataset where the combined duration of (reference speech + generated speech) is less than 20 seconds. i.e. 10 seconds of reference speech will generate 10 seconds of dubbed speech given the provided text.

**Comment by Plachtaa** [comment]
do it as you like then

**Comment by UniversalTechno** [comment]
I think the best method its to look first for a way to look if there a features or to modifies the existing great translation model like plachtaa mentionned MetaAI's SeamlessM4T. or others

**Comment by kdcyberdude** [comment]
Thank you, @Plachtaa and @UniversalTechno, for your suggestion. I'll take a look at the duration predictor component of M4T. 

I have a quick question, and I apologize if it sounds silly. Can you please confirm if the structure and shape of the x and y attention mask in the code above are correct for the intended objective?

**Comment by kdcyberdude** [comment]
Hi @Plachtaa, I'm currently facing a challenge and would appreciate your expertise in helping me understand why my loss is not converging. I'd like to provide some additional information so that you can better understand the situation. I'm using 40 seconds as the maximum duration for a batch in the DynamicBatchSampler(max_tokens), which is lower than the default value of 120 seconds. The model was trained on a single 4090 GPU. Additionally, I tried training it on A6000 (which has 48GB VRAM) for the last few epochs, with a longer maximum duration of 90 seconds per batch. I would be grateful for any insights or suggestions you may have on how to improve my model's performance. Thank you for your time and attention.

_[smoothing is 0]_


![Screenshot from 2023-10-27 23-39-28](https://github.com/Plachtaa/VALL-E-X/assets/34835322/d971efb0-1dd8-43dc-9f3f-0b206770b3e6)

![image](https://github.com/Plachtaa/VALL-E-X/assets/34835322/f0646a8e-daf7-43b6-baeb-df47a4712b6d)
